docs/DataProviders.md

```diff
@@ -650,4 +586,263 @@ const UserProfile = ({ userId }) => {
+          body: formData,
+        })
+        .then(({ json }) => ({ data: json }));
+    }
+
+    return baseDataProvider.create(resource, params);
+  },
+  update: (resource, params) => {
+    if (resource !== "posts") {
+      const formData = createPostFormData(params);
+      return fetchUtils
+        .fetchJson(`${endpoint}/${resource}`, {
+          method: "PUT",
+          body: formData,
+        })
+        .then(({ json }) => ({ data: json }));
+    }
+
+    return baseDataProvider.update(resource, params);
+  },
+};
+```
+
+### Sending Files To A Third-Party Service
+
+A common way to handle file uploads in single-page-apps is to first upload the file to a CDN such as [Cloudinary](https://cloudinary.com/) or [CloudImage](https://www.cloudimage.io/en/home), then use the the file URL generated by the CDN in the record creation / modification payload.
+
+Lets see an example with the [https://cloudinary.com/](Cloudinary) service, by adapting the `dataProvider` according to [their "Authenticated requests" example](hhtps://cloudinary.com/documentation/upload_images#authenticated_requests).
+
+To do that, you need an API that serves a [`signature`](https://cloudinary.com/documentation/upload_images#generating_authentication_signatures) the format that Cloudinary expect. To make it easier, you can install the [`cloudinary package`](https://cloudinary.com/documentation/node_integration#installation_and_setup).
+
+The following example generates and serves the signature needed to send files to Cloudinary. It uses Remix:
+
+```ts
+// handles the "get-cloudinary-signature" route and should be secured
+import { LoaderFunctionArgs, json } from "@remix-run/node";
+import cloudinary from "cloudinary";
+
+export const loader = ({ request }: LoaderFunctionArgs) => {
+  cloudinary.v2.config({
+    api_key: process.env.CLOUDINARY_API_KEY,
+    api_secret: process.env.CLOUDINARY_API_SECRET,
+    cloud_name: process.env.CLOUDINARY_CLOUD_NAME,
+    secure: false,
+  });
+
+  const timestamp = Math.round(new Date().getTime() / 1000);
+
+  const signature = cloudinary.v2.utils.api_sign_request(
+    {
+      timestamp: timestamp,
+    },
+    process.env.CLOUDINARY_API_SECRET as string
+  );
+
+  return json({
+    timestamp,
+    signature,
+    cloud_name: process.env.CLOUDINARY_CLOUD_NAME,
+    api_key: process.env.CLOUDINARY_API_KEY,
+  });
+};
+```
+
+The following `dataProvider` extends an existing one and leverages [`withLifecycleCallbacks`](#adding-lifecycle-callbacks) to modify the `dataProvider.create()` and `dataProvider.update()` methods for the `posts` resource only with the [`beforeSave`](./withLifecycleCallbacks.md#beforesave) methods. 
+
+It works as follow:
+- it grabs the Cloudinary signature to allow the autentication in the request;
+- it creates a new `FormData` object with the file received from the form;
+- it sends this file to the Cloudinary API; 
+- it hydrates the `params.picture` object with the data sent by Cloudinary;
```

Comment:
I'm not sure you can really use the term hydration here, as it holds a very [specific meaning](https://en.wikipedia.org/wiki/Hydration_(web_development)) in the web world. I'm worried it may be misleading here. Is there a reason I'm missing why you chose to use this word?

Reply:
To me, hydrating an object means taking an empty object, and filling it with data from a database, network, file system, or whatever. As `params.picture` was not empty, 'hydrate' might not fit in the end.

"Populates" could be OK for you?